### PR TITLE
port to C++ buildsystem

### PR DESCRIPTION
**Pull request by zingale**
This PR moves the unit test infrastructure to the C++ AMReX buildsystem.

At the moment, `test_eos` is done, and gives identical answers to the Fortran version

### CODE REVIEW

**Comment by zingale** [comment]
@mkatz @dwillcox happy for feedback on this conversion.  If you think it is okay, then I'll do `test_react` and the others, and then we can merge.

**Comment by maxpkatz** [comment]
Has this been tested against the Castro/Maestro test suites to ensure build compatibility?

File: Make.Microphysics [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+#------------------------------------------------------------------------------
+
+include $(AMREX_HOME)/Tools/GNUMake/Make.rules
+
+clean::
+	$(SILENT) $(RM) extern.F90
+	$(SILENT) $(RM) AMReX_buildInfo.cpp
+	$(RM) set_conserved.H set_indices.F90 state_sizes.f90
+
+# Older versions of CASTRO generated a "network.f90" file when using
```
Comment (maxpkatz) [inline Make.Microphysics]: Some of this stuff that is Castro-specific can probably be removed

Comment (zingale) [reply Make.Microphysics]: indeed

**Comment by zingale** [comment]
I'll run the Castro and Maestro suites with this branch now

File: unit_test/eos.F90 [hunk 2]
```diff
@@ -78,8 +77,8 @@ subroutine eos_init(small_temp, small_dens)
 
     if (present(small_temp)) then
        if (small_temp < mintemp) then
-          if (parallel_IOProcessor()) then
-             print *, 'EOS: small_temp cannot be less than the mintemp allowed by the EOS. Resetting small_temp to mintemp.'
+          if (amrex_pd_ioprocessor()) then
+             call bl_warning('EOS: small_temp cannot be less than the mintemp allowed by the EOS. Resetting small_temp to mintemp.')
```
Comment (maxpkatz) [inline unit_test/eos.F90]: I think print * is correct here, we don't want to continue using the deprecated bl_warning

Comment (zingale) [reply unit_test/eos.F90]: okay

File: unit_test/eos.F90 [hunk 3]
```diff
@@ -107,15 +106,31 @@ subroutine eos_init(small_temp, small_dens)
...
 
     !$acc routine seq
 
+    use eos_type_module, only: eos_t, composition
+#ifdef EXTRA_THERMO
+    use eos_type_module, only : composition_derivatives
+#endif
+    use actual_eos_module, only: actual_eos
+    use eos_override_module, only: eos_override
+#if (!(defined(AMREX_USE_CUDA) || defined(ACC)))
```
Comment (maxpkatz) [inline unit_test/eos.F90]: These can be replaced with the simpler

#ifdef AMREX_USE_GPU

Comment (maxpkatz) [reply unit_test/eos.F90]: #ifndef, in this case actually

File: unit_test/network.F90 [hunk 4]
```diff
@@ -89,10 +91,11 @@ function network_species_index(name) result(r)
 
   end function network_species_index
 
-  subroutine network_finalize
-    if (allocated(aion_inv)) then
-       deallocate(aion_inv)
-    end if
+  subroutine network_finalize()
+    implicit none
```
Comment (maxpkatz) [inline unit_test/network.F90]: I think we still want the deallocate(aion_inv) here, unless I'm missing something

File: unit_test/test_eos/main.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+{
+    amrex::Initialize(argc, argv);
+
+    main_main();
+
+    amrex::Finalize();
+    return 0;
+}
+
+void main_main ()
```
Comment (maxpkatz) [inline unit_test/test_eos/main.cpp]: What a missed opportunity to call this actual_main()

File: unit_test/network.F90 (hunk 4 above)
Comment (zingale) [reply unit_test/network.F90]: yes. Fixed now.  I think I copied these over from the Castro version.

File: unit_test/test_eos/main.cpp (hunk 5 above)
Comment (zingale) [reply unit_test/test_eos/main.cpp]: :fail:

File: unit_test/test_eos/variables.f90 [hunk 6]
```diff
@@ -31,13 +35,15 @@ module variables
 
      integer :: n_plot_comps = 0
 
-     character(len=20), allocatable :: names(:)
+     character(len=MAX_NAME_LEN), allocatable :: names(:)
```
Comment (maxpkatz) [inline unit_test/test_eos/variables.f90]: Is the get_next_plot_index stuff still relevant in the new build system?

Comment (zingale) [reply unit_test/test_eos/variables.f90]: yes... I use it inside the unit tests to define the variable and names for the plotfiles

File: unit_test/test_react/main.cpp [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+        do_react(AMREX_ARLIM_ARG(bx.loVect()), AMREX_ARLIM_ARG(bx.hiVect()),
+                 BL_TO_FORTRAN_ANYD(state[mfi]),
+                 &n_rhs_min, &n_rhs_max, &n_rhs_sum);
+
+    }
+
+
+    // get the name of the integrator
+    int nmodules = buildInfoGetNumModules();
```
Comment (maxpkatz) [inline unit_test/test_react/main.cpp]: Could you please describe what this little section does?

Comment (zingale) [reply unit_test/test_react/main.cpp]: the output files are of the format prefix.integrator, where integrator is `VODE`, `VBDF`, `BS`, `VODE90`.  These are passed into a python script at compile time that writes t he buildInfo C++ functions.  So now we query those functions, in particular, the ones that deal with the microphysics modules used to make the executable, to find the name of the integrator.

File: unit_test/test_react/main.cpp [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    std::cout << "number of modules = " << nmodules << std::endl;
```
Comment (zingale) [inline unit_test/test_react/main.cpp]: this is a debugging line that I can remove

File: unit_test/test_react/main.cpp (hunk 7 above)
Comment (maxpkatz) [reply unit_test/test_react/main.cpp]: Thanks!

**Comment by maxpkatz** [comment]
Can you please remove all of the GPackage.maks?

**Comment by maxpkatz** [comment]
> Can you please remove all of the GPackage.maks?
Nevermind, those are still needed by Maestro.

**Pull request merged**